tests/0056-balanced_group_mt.c

```diff
@@ -0,0 +1,329 @@
+                        rkqu = rd_kafka_queue_get_partition(rk,
+                                                            part.topic,
+                                                            part.partition);
+                        rd_kafka_queue_forward(rkqu, NULL);
+                        tids[part.partition] = spawn_thread(rkqu);
+                }
+
+                rebalanced = 1;
+
+                break;
+
+                case RD_KAFKA_RESP_ERR__REVOKE_PARTITIONS:
+                if (assign_cnt == 0)
+                        TEST_FAIL("asymetric rebalance_cb\n");
+                assign_cnt--;
+
+                for(int i = 0; i < partitions->cnt; ++i) {
+                        rd_kafka_topic_partition_t part = partitions->elems[i];
+                        rd_kafka_queue_t *rkqu;
+                        rkqu = rd_kafka_queue_get_partition(rk,
+                                                            part.topic,
+                                                            part.partition);
+                        rd_kafka_queue_forward(rkqu, NULL);
+                        tids[part.partition] = spawn_thread(rkqu);
+                }
+                rd_kafka_assign(rk, NULL);
+                break;
+
+                default:
+                TEST_FAIL("rebalance failed: %s\n",
+                          rd_kafka_err2str(err));
+                break;
+        }
+}
+
+
+static void get_assignment (rd_kafka_t *rk_c) {
+        while (!rebalanced) {
+                if (!rk_c)
+                        continue;
+                rd_kafka_message_t *rkmsg;
+                rkmsg = rd_kafka_consumer_poll(rk_c, 500);
+                if(rkmsg)
+                        rd_kafka_message_destroy(rkmsg);
+        }
+}
+
+int main_0056_balanced_group_mt (int argc, char **argv) {
+        const char *topic = test_mk_topic_name(__FUNCTION__, 1);
+        rd_kafka_t *rk_p, *rk_c;
+        rd_kafka_topic_t *rkt_p;
+        int msg_cnt = 1000;
+        int msg_base = 0;
+        int partition_cnt = 2;
+        exp_msg_cnt = msg_cnt*partition_cnt;
+        int partition;
+        uint64_t testid;
+        rd_kafka_topic_conf_t *default_topic_conf;
+        rd_kafka_topic_partition_list_t *topics;
+        rd_kafka_resp_err_t err;
+        test_timing_t t_assign, t_consume;
+        char errstr[512];
+
+        testid = test_id_generate();
+
+        /* Produce messages */
+        rk_p = test_create_producer();
+        rkt_p = test_create_producer_topic(rk_p, topic, NULL);
+
+        for (partition = 0 ; partition < partition_cnt ; partition++) {
+                test_produce_msgs(rk_p, rkt_p, testid, partition,
+                                  msg_base+(partition*msg_cnt), msg_cnt,
+                                  NULL, 0);
+        }
+
+        rd_kafka_topic_destroy(rkt_p);
+        rd_kafka_destroy(rk_p);
+
+        if (mtx_init(&lock, mtx_plain) != thrd_success)
+                TEST_FAIL("Cannot create mutex.\n");
```

Comment:
The TinyCThread documentation says that it can return thrd_error if the "request cannot be honored". No idea when that actually happens, but since it supposedly can happen, I checked for it.